File: flutter/lib/src/pylon_config.dart
import 'package:flutter/foundation.dart';

/// Configuration for the Pylon chat widget.
///
/// At minimum you need an [appId], which you can find under
/// Settings → Chat Widget in [app.usepylon.com](https://app.usepylon.com).
@immutable
class PylonConfig {
  /// Creates a configuration for the Pylon chat widget.
  const PylonConfig({
    required this.appId,
    this.enableLogging = true,
    this.primaryColor,
    this.debugMode = false,
    this.widgetBaseUrl,
    this.widgetScriptUrl,
    this.bubbleBottomOffset = 0,
  });

  /// The host the chat widget is served from when [widgetBaseUrl] is omitted.
  static const String defaultWidgetBaseUrl = 'https://widget.usepylon.com';

  /// Your Pylon app ID.
  final String appId;

  /// Whether the SDK writes its own diagnostics (and the widget's console
  /// output) to the platform log.
  final bool enableLogging;

  /// Overrides the widget's primary colour, as a CSS colour string.
  final String? primaryColor;

  /// Draws an overlay on top of the interactive regions the SDK tracks.
  ///
  /// Useful when a tap on the chat bubble is not registering: the overlay shows
  /// exactly which areas the SDK considers tappable.
  final bool debugMode;

  /// Overrides the host the widget is loaded from.
  ///
  /// Defaults to [defaultWidgetBaseUrl].
  final String? widgetBaseUrl;

  /// Overrides the URL of the widget script.
  ///
  /// Defaults to `<widgetBaseUrl>/widget/<appId>`.
  final String? widgetScriptUrl;

  /// Extra space, in logical pixels, kept clear below the chat bubble's
  /// resting position — so your app's own bottom chrome (a tab bar, a nav
  /// bar) doesn't sit under it.
  ///
  /// Applied on top of the device's own safe-area inset; zero (no extra
  /// space) by default. Has no effect on the full chat window, only the
  /// closed bubble.
  final double bubbleBottomOffset;

  /// Returns a copy of this config with the given fields replaced.
  PylonConfig copyWith({
    String? appId,
    bool? enableLogging,
    String? primaryColor,
    bool? debugMode,
    String? widgetBaseUrl,
    String? widgetScriptUrl,
    double? bubbleBottomOffset,
  }) {
    return PylonConfig(
      appId: appId ?? this.appId,
      enableLogging: enableLogging ?? this.enableLogging,
      primaryColor: primaryColor ?? this.primaryColor,
      debugMode: debugMode ?? this.debugMode,
      widgetBaseUrl: widgetBaseUrl ?? this.widgetBaseUrl,
      widgetScriptUrl: widgetScriptUrl ?? this.widgetScriptUrl,
      bubbleBottomOffset: bubbleBottomOffset ?? this.bubbleBottomOffset,
    );
  }

  /// Serialises this config for the platform channel.
  Map<String, Object?> toMap() {
    return <String, Object?>{
      'appId': appId,
      'enableLogging': enableLogging,
      'primaryColor': primaryColor,
      'debugMode': debugMode,
      'widgetBaseUrl': widgetBaseUrl,
      'widgetScriptUrl': widgetScriptUrl,
      'bubbleBottomOffset': bubbleBottomOffset,
    };
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is PylonConfig &&
        other.appId == appId &&
        other.enableLogging == enableLogging &&
        other.primaryColor == primaryColor &&
        other.debugMode == debugMode &&
        other.widgetBaseUrl == widgetBaseUrl &&
        other.widgetScriptUrl == widgetScriptUrl &&
        other.bubbleBottomOffset == bubbleBottomOffset;
  }

  @override
  int get hashCode => Object.hash(
    appId,
    enableLogging,
    primaryColor,
    debugMode,
    widgetBaseUrl,
    widgetScriptUrl,
    bubbleBottomOffset,
  );

  @override
  String toString() => 'PylonConfig(appId: $appId)';
}

File: android/pylon/src/main/java/com/pylon/chatwidget/PylonConfig.kt
package com.pylon.chatwidget

import java.net.URLEncoder

/**
 * Immutable configuration for the Pylon SDK. Use the [Builder] to customise
 * the settings and pass the result to [Pylon.initialize].
 */
data class PylonConfig internal constructor(
    val appId: String,
    val enableLogging: Boolean,
    val primaryColor: String?,
    val debugMode: Boolean,
    val widgetBaseUrl: String,
    val widgetScriptUrl: String,
    /**
     * Extra space, in device pixels, kept clear below the chat bubble's resting
     * position — so a host app's own bottom chrome (tab bar, nav bar) doesn't sit
     * under it. Applied on top of the safe-area inset; zero by default.
     */
    val bubbleBottomOffsetPx: Int
) {

    class Builder internal constructor(private val appId: String) {
        var enableLogging: Boolean = true
        var primaryColor: String? = null
        var debugMode: Boolean = false
        var widgetBaseUrl: String = DEFAULT_WIDGET_BASE_URL
        var widgetScriptUrl: String = defaultScriptUrl(appId)
        var bubbleBottomOffsetPx: Int = 0

        internal fun build(): PylonConfig {
            val scriptUrl = widgetScriptUrl.ifBlank { defaultScriptUrl(appId) }
            return PylonConfig(
                appId = appId,
                enableLogging = enableLogging,
                primaryColor = primaryColor,
                debugMode = debugMode,
                widgetBaseUrl = widgetBaseUrl.ifBlank { DEFAULT_WIDGET_BASE_URL },
                widgetScriptUrl = scriptUrl,
                bubbleBottomOffsetPx = bubbleBottomOffsetPx
            )
        }
    }

    companion object {
        private const val DEFAULT_WIDGET_BASE_URL = "https://widget.usepylon.com"

        private fun defaultScriptUrl(appId: String): String {
            val encodedAppId = URLEncoder.encode(appId, "UTF-8")
            return "$DEFAULT_WIDGET_BASE_URL/widget/$encodedAppId"
        }

        fun build(appId: String, block: Builder.() -> Unit = {}): PylonConfig {
            return Builder(appId).apply(block).build()
        }

        fun from(existing: PylonConfig, block: Builder.() -> Unit): PylonConfig {
            val builder = Builder(existing.appId).apply {
                enableLogging = existing.enableLogging
                primaryColor = existing.primaryColor
                debugMode = existing.debugMode
                widgetBaseUrl = existing.widgetBaseUrl
                widgetScriptUrl = existing.widgetScriptUrl
                bubbleBottomOffsetPx = existing.bubbleBottomOffsetPx
            }
            builder.block()
            return builder.build()
        }
    }
}
